rotkehlchen/data_import/utils.py

```diff
@@ -0,0 +1,71 @@
+from abc import ABCMeta, abstractmethod
+from pathlib import Path
+from typing import Any, List, Optional, Tuple
+
+from rotkehlchen.accounting.ledger_actions import LedgerAction
+from rotkehlchen.db.dbhandler import DBHandler
+from rotkehlchen.db.drivers.gevent import DBCursor
+from rotkehlchen.db.ledger_actions import DBLedgerActions
+from rotkehlchen.errors.misc import InputError
+from rotkehlchen.exchanges.data_structures import AssetMovement, Trade
+
+ROWS_PER_CONTEXT_SWITCH = 500
+
+ITEMS_PER_DB_WRITE = 200
+
+
+class BaseExchangeImporter(metaclass=ABCMeta):
+    def __init__(self, db: DBHandler) -> None:
+        self.db = db
+        self.db_ledger = DBLedgerActions(self.db, self.db.msg_aggregator)
+        self._trades: List[Trade] = []
+        self._asset_movements: List[AssetMovement] = []
+        self._ledger_actions: List[LedgerAction] = []
+        self.cursor: Optional[DBCursor] = None
+
+    def import_csv(self, filepath: Path, **kwargs: Any) -> Tuple[bool, str]:
+        try:
+            with self.db.user_write() as cursor:
+                self.cursor = cursor
+                self._import_csv(filepath=filepath, **kwargs)
+                self.flush_all()
+            return True, ''
+        except InputError as e:
+            return False, str(e)
+
+    @abstractmethod
+    def _import_csv(self, filepath: Path, **kwargs: Any) -> None:
+        """The method that processes csv. Should be implemented by subclasses.
+        May raise:
+        - InputError if one of the rows is malformed
+        """
+        ...
+
+    def add_trade(self, trade: Trade) -> None:
+        self._trades.append(trade)
+        self.maybe_flush_all()
+
+    def add_asset_movement(self, asset_movement: AssetMovement) -> None:
+        self._asset_movements.append(asset_movement)
+        self.maybe_flush_all()
+
+    def add_ledger_action(self, ledger_action: LedgerAction) -> None:
+        self._ledger_actions.append(ledger_action)
+        self.maybe_flush_all()
+
+    def maybe_flush_all(self) -> None:
+        if len(self._trades) + len(self._asset_movements) + len(self._ledger_actions) >= ITEMS_PER_DB_WRITE:  # noqa: E501
+            self.flush_all()
```

Comment:
Hey @nebolax so I checked this again and I am not sure we are on the same page.

I asked in discord: 

>1. Keep the flushing mechanism but make sure no commits happen until the very end and if any error is seen everything is rolled back or if the app shuts down nothing is commited. 

You said you would do that but as I understand the code right now will write every `ITEMS_PER_DB_WRITE` to the DB and commit. notice the context manager there.

What I meant is to commit at the very end once finished. And if not finished do not commit and rollback everything in case of an exception.

Does this make sense?